### PR TITLE
Reduce memory usage when returning large result sets

### PR DESCRIPTION
**Pull request by Azquelt**
When returning large resultsets, stream assets one by one from the database to the response without holding the entire response in memory.

### CODE REVIEW

File: server/src/main/java/com/ibm/ws/lars/rest/PersistenceBean.java [hunk 1]
```diff
@@ -262,27 +269,24 @@ private BasicDBObject createFilterObject(String field, Condition condition) {
...
-                convertObjectIdToHexString(obj);
-                results.add(obj);
-            }
+        if (sortObject != null) {
+            cursor.sort(sortObject);
         }
-        return results;
+
+        AssetCursor result = new MongoAssetCursor(cursor);
+        result.addOperation(CONVERT_OID_TO_HEX);
```
Comment (idlewis) [inline server/src/main/java/com/ibm/ws/lars/rest/PersistenceBean.java:292]: Both places we currently the new MongoAssetCursor, we add the convert operation. would it make sense to have a factory for this? MongoAssetCursor.newCursorWithIdConversion(cursor) or similar?

File: server/src/test/java/com/ibm/ws/lars/rest/BasicAssetCursor.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return assets.hasNext();
+    }
+
+    @Override
+    public Asset next() {
+        return new Asset(assets.next());
+    }
+
+    @Override
+    public void remove() {
```
Comment (idlewis) [inline server/src/test/java/com/ibm/ws/lars/rest/BasicAssetCursor.java:51]: Don't think there is any use-case for this, but isn't 'remove' supported by the map iterator?

File: server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * returned.
+ */
+@JsonSerialize(as = Iterator.class)
+public interface AssetCursor extends Iterator<Asset> {
+    /**
+     * Get the total number of assets which will be returned from this cursor
+     *
+     * @return the number of assets
+     */
+    public int getCount();
```
Comment (idlewis) [inline server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java]: any reason why you called this getCount? It seems to replace a method call to 'size' in lots of places, and wraps a call to 'size' in the case of the mongo cursor.

File: server/src/main/java/com/ibm/ws/lars/rest/AssetCursorWriter.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+        return AssetCursor.class.isAssignableFrom(type) && mediaType.isCompatible(MediaType.APPLICATION_JSON_TYPE);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void writeTo(AssetCursor cursor, Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType, MultivaluedMap<String, Object> httpHeaders,
```
Comment (idlewis) [inline server/src/main/java/com/ibm/ws/lars/rest/AssetCursorWriter.java:53]: You're not checking the mediatype here. At this point, do we just have to assume that jax-rs has called the above method and is giving us something sensible?

Comment (Azquelt) [reply server/src/main/java/com/ibm/ws/lars/rest/AssetCursorWriter.java:53]: Yeah, that was what I was assuming

File: server/src/test/java/com/ibm/ws/lars/rest/BasicAssetCursor.java (hunk 2 above)
Comment (Azquelt) [reply server/src/test/java/com/ibm/ws/lars/rest/BasicAssetCursor.java:51]: It is, but it's not supported in the MongoAssetCursor, so there doesn't seem much point implementing it in the test implementation

File: server/src/main/java/com/ibm/ws/lars/rest/PersistenceBean.java (hunk 1 above)
Comment (Azquelt) [reply server/src/main/java/com/ibm/ws/lars/rest/PersistenceBean.java:292]: Hmm, possibly? Or I could put this functionality into `MongoAssetCursor` itself on the basis that we will always need to convert the id when turning a MongoDB record into an asset.

On the other hand, it fits quite nicely into an operation.

File: server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java (hunk 3 above)
Comment (Azquelt) [reply server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java]: Not really, `getCount()` made sense to me when writing it, but you're correct that other objects use `size()`

Comment (Azquelt) [reply server/src/main/java/com/ibm/ws/lars/rest/model/AssetCursor.java]: For added confusion, `DBCursor` has `length()`, `count()` and `size()` which all do different things

**Comment by Azquelt** [comment]
I checked and we do have a check to ensure the search match score is not returned in the response. It's at the end of `PersistenceBeanTest.testRetrieveAllFiltered()`

**Review by idlewis — APPROVED** [review]
Ok

**Pull request merged**